File: src/android/IntuneUPN.java
/*
 * Copyright 2022 Zalaris ASA.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.zalaris.cordova.intuneupn;

import org.apache.cordova.CordovaPlugin;
import org.apache.cordova.CallbackContext;
import org.json.JSONObject;
import org.json.JSONArray;
import org.json.JSONException;
import android.util.Log;
import java.lang.reflect.InvocationTargetException;
import java.lang.reflect.Method;
import java.lang.reflect.Field;

public class IntuneUPN extends CordovaPlugin {
    private final String TAG = "IntuneUPNPlugin";

    /**
     * Executes the request.
     *
     * This method is called from the WebView thread. To do a non-trivial amount of
     * work, use:
     * cordova.getThreadPool().execute(runnable);
     *
     * To run on the UI thread, use:
     * cordova.getActivity().runOnUiThread(runnable);
     *
     * @param action          The action to execute.
     * @param args         The exec() arguments in JSON form.
     * @param callbackContext The callback context used when calling back into
     *                        JavaScript.
     * @return Whether the action was valid.
     */
    @Override
    public boolean execute(String action, JSONArray args, CallbackContext callbackContext) {
        if ("getUPN".equals(action)) {
            try {
                JSONObject upn = this.getUPN(callbackContext);
                callbackContext.success(upn);
                return true;

            } catch (Exception e){
                Log.e(TAG, e.getMessage());

                callbackContext.error("Something wrong has happened! Most likely the app is not wrapped using MS Intune Wrapping Tool or does not implement MS Intune SDK.");
                return false;
            }
        } else {
            callbackContext.error("Wrong method!");
            return false;
        }
    }

    private JSONObject getUPN(CallbackContext callbackContext) throws Exception {
        JSONObject result = new JSONObject();

        try {
            Class clsMamComponents = Class.forName("com.microsoft.intune.mam.client.app.MAMComponents");
            Class clsUserInfo = Class.forName("com.microsoft.intune.mam.policy.MAMUserInfo");
            Class clsMAMIdentityManager = Class.forName("com.microsoft.intune.mam.client.identity.MAMIdentityManager");

            // Reflect MAMComponents.get
            Method methodMamGet = clsMamComponents.getMethod("get", new Class[] { Class.class });

            // Reflect MAMComponents.get(MAMUserInfo.class)
            Object objUserInfo = methodMamGet.invoke(clsMamComponents, clsUserInfo);

            // Get User's UPN
            Method methodGetUser = objUserInfo.getClass().getMethod("getPrimaryUser", null);
            String userUPN = methodGetUser.invoke(objUserInfo).toString();
            result.put("upn", userUPN);

            // Reflect MAMComponents.get(MAMIdentityManager.class)
            Object objIdentityManager = methodMamGet.invoke(clsMamComponents, clsMAMIdentityManager);

            // Get User Identity instance (contains Azure Tenant ID)
            Method methodFromString = objIdentityManager.getClass().getMethod("fromString", new Class[] { String.class });
            Object userIdentity = methodFromString.invoke(objIdentityManager, userUPN);

            // Get Azure Tenant ID
            try {
                Method methodGetTenantId = userIdentity.getClass().getMethod("tenantId", null);
                String tenantId = methodGetTenantId.invoke(userIdentity).toString();
                result.put("tenantId", tenantId);

            } catch (NoSuchMethodException e) {
                Log.i(TAG, "There is no method " + e.getMessage() + ". Switching to reflect the field mTenantAadId.");

                Field fTenantId = userIdentity.getClass().getDeclaredField("mTenantAadId");
                fTenantId.setAccessible(true);
                String tenantId = fTenantId.get(userIdentity).toString();
                result.put("tenantId", tenantId);
            }

        } catch (NoSuchMethodException e) {
            throw new Exception("There is no method " + e.getMessage());
        } catch (InvocationTargetException e) {
            throw new Exception("Cannot execute method " + e.getMessage());
        } catch (IllegalAccessException e) {
            throw new Exception("IllegalAccessException: " + e.getMessage());
        } catch (ClassNotFoundException e) {
            throw new Exception("There is no class " + e.getMessage());
        } catch (JSONException e) {
            throw new Exception("JSONException: " + e.getMessage());
        } catch (NoSuchFieldException e) {
            throw new Exception("There is no field " + e.getMessage());
        }

        return result;
    }
}

File: src/ios/IntuneUPN.swift
/*
 * Copyright 2022 Zalaris ASA.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
 
import Foundation

extension String: Error {} // Enables you to throw a string

extension String: LocalizedError { // Adds error.localizedDescription to Error instances
    public var errorDescription: String? { return self }
}

@objc(IntuneUPN) class IntuneUPN : CDVPlugin {
    
    @objc(getUPN:)
    func getUPN(command: CDVInvokedUrlCommand) {
        var pluginResult = CDVPluginResult(
            status: CDVCommandStatus_ERROR,
            messageAs: "Something wrong happend! Most likely the app is not wrapped using Intune Wrapping Tool or does not implement MS Intune SDK and/or MSAL."
        )
        
        struct PluginOutput: Encodable {
            let upn: String
            let tenantId: String
        }

        //Get Client ID from settings, set default in case of failure
        var clientId = self.commandDelegate!.settings["clientid"] as? String ?? "6c7e8096-f593-4d72-807f-a5f86dcc9c77"

        //Overwrite Client ID if passed to the plugin explicitly
        if let customClientId = command.argument(at: 0) as? String {
           clientId = customClientId
        }
        
        do {
            //Get UPN
            guard let clsIntuneMAMEnrollmentManager = NSClassFromString("IntuneMAMEnrollmentManager") as? NSObject.Type else { throw "Missing class IntuneMAMEnrollmentManager" }
            let objInstance = clsIntuneMAMEnrollmentManager.perform(NSSelectorFromString("instance")).takeUnretainedValue()
            let objEnrolledAccount = objInstance.perform(NSSelectorFromString("enrolledAccount")).takeUnretainedValue()
            let upn: String = objEnrolledAccount as? String ?? ""
            NSLog(upn)
            
            //Get Tenant Id           
            guard let clsMSALPublicClientApp = NSClassFromString("MSALPublicClientApplication") as? NSObject.Type else { throw "Missing class MSALPublicClientApplication" }
            let objMSALPublicClientApp = clsMSALPublicClientApp.init()
            let objMSALClientAppInitialized = objMSALPublicClientApp.perform(NSSelectorFromString("initWithClientId:error:"), with: clientId, with: nil)?.takeUnretainedValue()
            
            let objMSALAccount = objMSALClientAppInitialized?.perform(NSSelectorFromString("accountForUsername:error:"), with: upn, with: nil)?.takeUnretainedValue()
            
            let homeAccountId = objMSALAccount?.value(forKey: "homeAccountId")
            let tenantId: String = (homeAccountId as? AnyObject)?.value(forKey: "tenantId") as? String ?? ""
            NSLog(tenantId)
            
            //Build plugin output
            let pluginOutput = PluginOutput(upn: upn, tenantId: tenantId)
            let jsonEncoder = JSONEncoder()

            let encodePluginOutputData = try jsonEncoder.encode(pluginOutput)
            let encodePluginOutput = String(data: encodePluginOutputData, encoding: .utf8)!
        
            pluginResult = CDVPluginResult(
                status: CDVCommandStatus_OK,
                messageAs: encodePluginOutput
            )
               
            self.commandDelegate!.send(
                pluginResult,
                callbackId: command.callbackId
            )
            
        } catch {
            NSLog(error.localizedDescription)
            
            self.commandDelegate!.send(
                pluginResult,
                callbackId: command.callbackId
            )
        }
    }
}